tests/e2e/e2e_test.go

```diff
@@ -171,16 +173,61 @@ func (s *IntegrationTestSuite) TestTWAP() {
+	s.T().Log("querying for the second TWAP to now before swap, must equal to first")
+	twapFromBeforeSwapToBeforeSwapTwo, err := chainANode.QueryGetArithmeticTwapToNow(poolId, denomOne, denomTwo, timeBeforeSwap.Add(50*time.Millisecond))
+	s.Require().NoError(err)
+
+	// Since there were no swaps between the two queries, the TWAPs should be the same.
+	s.Require().Equal(twapFromBeforeSwapToBeforeSwapOne, twapFromBeforeSwapToBeforeSwapTwo)
+
+	s.T().Log("performing swap")
+	chainANode.SwapExactAmountIn(coinIn, minAmountOut, fmt.Sprintf("%d", poolId), denomTwo, swapWalletAddr)
 
 	keepPeriodCountDown := time.NewTimer(initialization.TWAPPruningKeepPeriod)
 
-	// Make sure still producing blocks.
-	chainA.WaitUntilHeight(heightBeforeSwap + 3)
+	// Make sure that we are still producing blocks and move far enough for the swap TWAP record to be created
+	// so that we can measure start time post-swap (timeAfterSwap).
+	chainA.WaitForNumHeights(2)
+
+	// Measure time after swap and wait for a few blocks to be produced.
+	// This is needed to ensure that start time is before the block time
+	// when we query for TWAP.
+	timeAfterSwap := chainANode.QueryLatestBlockTime()
+	chainA.WaitForNumHeights(2)
+
+	// TWAP "from before to after swap" should be different from "from before to before swap"
+	// because swap introduces a new record with a different spot price.
+	s.T().Log("querying for the TWAP from before swap to now after swap, must not equal to TWAPs before swap")
+	twapFromBeforeSwapToAfterSwap, err := chainANode.QueryGetArithmeticTwapToNow(poolId, denomOne, denomTwo, timeBeforeSwap)
+	s.Require().NoError(err)
+	s.Require().NotEqual(twapFromBeforeSwapToBeforeSwapOne, twapFromBeforeSwapToAfterSwap)
+
+	// TWAP "from after to after swap" should be different from "from before to after swap"
+	// because the former has a higher time weight for the after swap period.
+	s.T().Log("querying for the TWAP from after swap to now, must not equal to TWAPs before swap")
+	twapFromAfterToNow, err := chainANode.QueryGetArithmeticTwapToNow(poolId, denomOne, denomTwo, timeAfterSwap)
+	s.Require().NoError(err)
+	s.Require().NotEqual(twapFromBeforeSwapToAfterSwap, twapFromAfterToNow)
+
+	s.T().Log("querying for the TWAP from after swap to after swap + 10ms, must not equal to TWAPs before swap")
+	twapFromAfterToAfterSwapAndBeforePruning, err := chainANode.QueryGetArithmeticTwap(poolId, denomOne, denomTwo, timeAfterSwap, timeAfterSwap.Add(10*time.Millisecond))
+	s.Require().NoError(err)
+	s.Require().NotEqual(twapFromBeforeSwapToAfterSwap, twapFromAfterToAfterSwapAndBeforePruning)
+
+	// These must be equal because they are calculated over time ranges with the stable and equal spot price.
+	// There are potential rounding errors requiring us to approximate the comparison.
+	osmoassert.DecApproxEq(s.T(), twapFromAfterToAfterSwapAndBeforePruning, twapFromAfterToNow, sdk.NewDecWithPrec(1, 3))
```

Comment:
Note to reviewers:

I've been debugging this with extra logs added.

I observed that records are picked up from the correct heights and with the correct spot prices.

I also logged time deltas and accumulator differences from:
https://github.com/osmosis-labs/osmosis/blob/7a2db79f406adfd3a5099e4acb1b1936b6e25c06/x/twap/types/utils.go#L55-L56

It seems that these are not exactly equal in cases when the time deltas are different. So I suspect that these are rounding issues caused by truncations.

Another potential source is sig fig round for spot prices. I attempted to increase precision by 4 sig figs for testing purposes. Although it reduced the delta, it did not completely eliminate it.